src/CommitmentsV2/SFA.DAS.CommitmentsV2.MessageHandlers/EventHandlers/RecordedAct1CompletionPaymentEventHandler.cs

```diff
@@ -0,0 +1,53 @@
+﻿using Microsoft.Extensions.Logging;
+using NServiceBus;
+using SFA.DAS.CommitmentsV2.Messages.Events;
+using System;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using SFA.DAS.CommitmentsV2.Data;
+using SFA.DAS.CommitmentsV2.Domain.Interfaces;
+using SFA.DAS.CommitmentsV2.Types;
+
+namespace SFA.DAS.CommitmentsV2.MessageHandlers.EventHandlers
+{
+    public class RecordedAct1CompletionPaymentEventHandler : IHandleMessages<RecordedAct1CompletionPaymentFakeEvent>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _dbContext;
+        private readonly ILogger<RecordedAct1CompletionPaymentEventHandler> _logger;
+
+        public RecordedAct1CompletionPaymentEventHandler(Lazy<ProviderCommitmentsDbContext> dbContext, ILogger<RecordedAct1CompletionPaymentEventHandler> logger)
+        {
+            _dbContext = dbContext;
+            _logger = logger;
+        }
+
+        public async Task Handle(RecordedAct1CompletionPaymentFakeEvent message, IMessageHandlerContext context)
+        {
+            try
+            {
+                if (message.ApprenticeshipId.HasValue)
+                {
+                    var apprentice = await _dbContext.Value.Apprenticeships.Include(x=>x.Cohort).SingleAsync(x => x.Id == message.ApprenticeshipId);
+
+                    switch (apprentice.Status)
+                    {
+                        case ApprenticeshipStatus.Live:
+                            apprentice.Complete(message.EventTime.DateTime);
+                            break;
+                        case ApprenticeshipStatus.Completed:
+                            apprentice.UpdateCompletionDate(message.EventTime.DateTime);
+                            break;
+                        default:
+                            _logger.LogWarning($"Warning {nameof(RecordedAct1CompletionPaymentEventHandler)} - Cannot process CompletionEvent for apprenticeshipId {apprentice.Id} as status is {apprentice.Status}");
+                            break;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Error processing RecordedAct1CompletionPaymentEvent", e);
+                throw;
```

Comment:
Does the catch and rethrow mean that, in the case of some permanent errors we are going to retry and error-queue the message ultimately, when it would be better to log as an error supress?

Reply:
Good point, these would appear as noise for retries, but if we don't do this then we only get visibility of the error from the poisoned message. Which we have to ask for (or has that changed, since we now have our own error queue). Not sure what you mean by error suppress

Reply:
As an example, right here in this handler, if the apprenticeship isn't either Live (in which case that Status field _is_ of use) or Paused then we should log a Warning and return without exception. This is what I meant by suppress any error - avoid the throw and therefore retry. Hope that makes sense!

> In addition to the Invariant check, we should handle certain invalid states in the event handler itself, so that we can avoid an exception and therefore a retry for a permanent failure:
> 
> On receipt of an event, if the Apprenticeship is not either Live or Complete (it is Stopped, Paused, or Waiting to Start) then log a WARNing with a clear message and return without performing any state change

Reply:
We do this as the default action on the switch statement